### PR TITLE
Rl

### PR DESCRIPTION
**Pull request by zikun-li**
Changes:
- `CMakeList.txt`: compile mode changed to `release`, optimization level changed from `o2` to `o3`.
- Added benchmarks circuits `circuit/QFT_and_Adders/`.
- Reoranized benchmark qasm files. In `circuits`, added  1)`nam_circs`, containing benchmarks in nam's gate set; 2) `nam_rm_circs`, containing benchmarks in nam's gate set after rotation merging; 3) `t_tdg_circs`, containing benchmarks in t_tdg gate set; 4) `t_tdg_rm_circs`, containing benchmarks in t_tdg gate set after rotation merging. All files in these four directories are originated from `nam-benchmark`, and used greedy toffoli flip to decompose toffoli gates.
- Added four python test files in `src/test`, which are 1) `test_nam_rm.py` and `test_nam_worm.py`, testing optimization in nam's gate set with and without rotation merging respectively; 2) `test_t_tdg_rm.py` and `test_t_tdg_worm.py` testing optimization in t_tdg gate set with and without rotation merging respectively.
- Added new `optimize` API in `Graph`.
- Added `src/test/test_optimize` to test the new `optimize` API.
- Added API to directly load a `qasm` file into `Graph` in `Graph`.
- Added `src/test/test_from_and_to_qasm` to test the new API that load `qasm` to `Graph`.
- Added constrain for `GraphXfer` creation, that is, the source circuit must be connected.
- Formatted some unformatted qasm files.

### CODE REVIEW

**Comment by zikun-li** [comment]
> It seems that there are some conflicts with the master branch...

I'll try to resolve them right away.

File: src/quartz/tasograph/tasograph.cpp [hunk 1]
```diff
@@ -1661,7 +1746,9 @@ Graph::toffoli_flip_by_instruction(GateType target_rotation, GraphXfer *xfer,
                                    std::vector<int> instruction) {
   std::shared_ptr<Graph> graph(new Graph(*this));
   std::shared_ptr<Graph> new_graph(nullptr);
+  int a = 0;
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1749]: Remove this debug output?

File: src/quartz/tasograph/tasograph.cpp [hunk 2]
```diff
@@ -1583,6 +1600,74 @@ std::shared_ptr<Graph> Graph::optimize(
...
+                                       int gate_count_upper_bound,
+                                       std::string circuit_name,
+                                       bool print_message, int timeout) {
+  auto start = std::chrono::steady_clock::now();
+  std::priority_queue<std::shared_ptr<Graph>,
+                      std::vector<std::shared_ptr<Graph>>, GraphCompare>
+      candidates;
+  std::set<size_t> hashmap;
+  std::shared_ptr<Graph> best_graph(new Graph(*this));
+  auto best_gate_cnt = gate_count();
```
Comment (xumingkuan) [inline src/quartz/tasograph/tasograph.cpp:1613]: We want to support other cost functions. It's better to use `total_cost()` here (also change the variable name and the message printed).

**Review by xumingkuan — COMMENTED** [review]
Thanks for the work!

I think we should use `total_cost` in `optimize` in case users want to change the cost function.

File: src/quartz/tasograph/tasograph.cpp (hunk 1 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp]: Done.

File: src/quartz/tasograph/tasograph.cpp (hunk 2 above)
Comment (zikun-li) [reply src/quartz/tasograph/tasograph.cpp]: Done.

**Review by xumingkuan — APPROVED** [review]
LGTM!

**Pull request merged**